Source/WebCore/html/ImageDataArray.h

```diff
@@ -28,34 +28,147 @@
+template <> struct TypedArrayForStorageFormat<ImageDataStorageFormat::Uint8> {
+    using TypedArray = JSC::Uint8ClampedArray;
+};
+template <> struct TypedArrayForStorageFormat<ImageDataStorageFormat::Float16> {
+    using TypedArray = JSC::Float16Array;
+};
+
 class ImageDataArray {
 public:
-    // FIXME: This API is the minimum subset as currently used by ImageData. More to come...
+    static std::optional<ImageDataArray> tryCreateUninitialized(size_t length, ImageDataStorageFormat storageFormat)
+    {
+        return tryCreate(length, storageFormat, false);
+    }
+
+    static std::optional<ImageDataArray> tryCreateZeroFilled(size_t length, ImageDataStorageFormat storageFormat)
+    {
+        return tryCreate(length, storageFormat, true);
+    }
+
+    static constexpr bool isSupported(JSC::TypedArrayType typedArrayType)
+    {
+        switch (typedArrayType) {
+        case JSC::TypeUint8Clamped:
+        case JSC::TypeFloat16:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    static constexpr bool isSupported(const JSC::ArrayBufferView& arrayBufferView)
+    {
+        return isSupported(arrayBufferView.getType());
+    }
+
+    static std::optional<ImageDataArray> tryCreate(Ref<JSC::ArrayBufferView>&& arrayBufferView)
+    {
+        if (isSupported(arrayBufferView))
+            return { ImageDataArray(WTFMove(arrayBufferView)) };
+        return { };
+    }
+
     ImageDataArray(Ref<JSC::Uint8ClampedArray>&& data)
-        : m_bufferView(WTFMove(data))
+        : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
     { }
 
+    ImageDataArray(Ref<JSC::Float16Array>&& data)
+        : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+    { }
+
+    ImageDataStorageFormat storageFormat() const
+    {
+        switch (m_bufferView->getType()) {
+        case JSC::TypeUint8Clamped: return ImageDataStorageFormat::Uint8;
+        case JSC::TypeFloat16: return ImageDataStorageFormat::Float16;
+        default: RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+        }
+    }
+
+    template <typename F>
+    static auto visit(JSC::ArrayBufferView& bufferView, F&& f)
+    {
+        // Always try Uint8ClampedArray first, as it should be the most frequent.
+        if (auto* array = dynamicDowncast<JSC::Uint8ClampedArray>(bufferView))
+            return std::forward<F>(f)(*array);
+        if (auto* array = dynamicDowncast<JSC::Float16Array>(bufferView))
+            return std::forward<F>(f)(*array);
+        RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+    }
+
+    template <typename F>
+    auto visit(F&& f) const
+    {
+        return visit(m_bufferView.get(), std::forward<F>(f));
+    }
+
+    Ref<JSC::Uint8ClampedArray> asUint8ClampedArray() const;
+    Ref<JSC::Float16Array> asFloat16Array() const;
+
+    JSC::ArrayBufferView& getArrayBufferView() const { return m_bufferView.get(); }
```

Comment:
Can't `getArrayBufferView()` be just `data()`?

Reply:
I guess I was tired of all the `data` functions, giving us expressions like `blah->data()->data()->data()` 😄
Anyway, I've just removed this one, it was only used once in this same file, so no real need.

(And thanks for the thorough review, I'm working on addressing your other more important comments now...)

Reply:
Trying to remove the `visit` in JSImageDataCustom.cpp (as you suggested), I now need this getter-returning-ArrayBufferView back!
So the caller would be `toJS(..., arrayBufferView())` (I've removed the 'get').

Are you sure you want "data"? It would make the caller look like: `toJS(..., data.data())`.
Anyway, I'll try to rework some names for clarity in that JSImageDataCustom.cpp, please have a look when I post the update...